Add GitHub Action to Automatically Publish Release Tarballs

felixhandte opened the pull request:
This PR introduces a GitHub action that is triggered on release creation, which creates the release tarball, compresses it, hashes it, signs it, and attaches all of those files to the release.

Here's an example invocation on my repo:

![Screenshot from 2021-03-12 17-33-48](https://user-images.githubusercontent.com/1154694/111005657-35a0f100-8359-11eb-9469-e72d4c26250c.png)

I wanted to also generate the windows artifacts but windows is hard and I gave up.

Because this PR signs the artifacts when a key is provided, this is one step towards addressing #2520.

felixhandte committed:
Add GitHub Action to Automatically Publish Release Tarballs

This commit introduces a GitHub action that is triggered on release creation,
which creates the release tarball, compresses it, hashes it, signs it, and
attaches all of those files to the release.

felixhandte committed:
Remove CircleCI Artifact Generation

terrelln commented:
This looks good to me! I don't have a strong preference on the tarball names. As long as we aren't changing the old names, it shouldn't be a huge deal to add `v`. But, I guess I'd very slightly prefer keeping `zstd-1.4.9`.

felixhandte committed:
Maintain Artifact Name Backwards Compatibility

When the tag is `v1.2.3`, name the artifacts `zstd-1.2.3.tar*` rather than
`zstd-v1.2.3.tar*`. When the tag doesn't match, use the full tag.

felixhandte committed:
Allow a Passphrase on the Key

mohe2015 commented:
I think this is fine as an MVP but I really don't like it that this seems to do signing on Github's servers. The keys should be guarded as much as possible - some people even use hardware security keys.
